### PR TITLE
Refine RL workflow & tune RL models under GYM

### PR DESCRIPTION
**Pull request by lihuoran**
# Description

- Refine RL workflow
  - Refine rollout workflow. Now we can:
    - Run specific number of steps in rollout.
    - Use `num_eval_episodes` to control the number of episodes during evaluation.
  - Add `AbsEnvSampler.metrics`. Let env samplers use this attribute to manage metrics during rollout.
  - Add `Callback`, a generate interface to add customized operations in each phase of the workflow. Add two `Callback` instances: `Checkpoint` & `MetricsRecorder`.
  - Other miner code refinements.
- Tune RL models under GYM
  - Add DDPG. Optimize the performance of PPO, SAC, and DDPG.
  - Fix several RL algorithm bugs.
  - Re-organize RL jobs' output paths.
  - Re-organize RL test's file structure.

## Linked issue(s)/Pull request(s)

- [issue_number](issue_link)

## Type of Change

- [ ] Non-breaking bug fix
- [ ] Breaking bug fix
- [ ] New feature
- [ ] Test
- [ ] Doc update
- [ ] Docker update

## Related Component

- [ ] Simulation toolkit
- [ ] RL toolkit
- [ ] Distributed toolkit

## Has Been Tested

- OS:
  - [ ] Windows
  - [ ] Mac OS
  - [ ] Linux
- Python version:
  - [ ] 3.7
  - [ ] 3.8
  - [ ] 3.9
- Key information snapshot(s):

## Needs Follow Up Actions

- [ ] New release package
- [ ] New docker image

## Checklist

- [ ] Add/update the related comments
- [ ] Add/update the related tests
- [ ] Add/update the related documentations
- [ ] Update the dependent downstream modules usage

### CODE REVIEW

File: maro/rl/workflows/callback.py [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+    def on_validation_end(
+        self,
+        env_sampler: EnvSampler,
+        training_manager: TrainingManager,
+        logger: LoggerV2,
+        ep: int,
+    ) -> None:
+        pass
+
+    def on_test_start(
```
[the remark this answers is not in the document]
Comment (lihuoran) [reply maro/rl/workflows/callback.py:71]: Theoretically, validation and test could be different. It is just for now that we reuse validation to do testing (actually we do not really have a "testing" phase). I personally think it is okay to leave these methods for potential usage.

File: maro/rl/workflows/callback.py [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+            else:
+                self._full_metrics[ep] = metrics
+            if ep in self._valid_metrics:
+                self._valid_metrics[ep].update(metrics)
+            else:
+                self._valid_metrics[ep] = metrics
+        self._dump_metric_history()
+
+
+SUPPORTED_CALLBACK_FUNC = {
```
[the remark this answers is not in the document]
Comment (lihuoran) [reply maro/rl/workflows/callback.py:164]: Will do.

File: maro/rl/workflows/callback.py [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+        self,
+        env_sampler: EnvSampler,
+        training_manager: TrainingManager,
+        logger: LoggerV2,
+        ep: int,
+    ) -> None:
+        pass
+
+
+class Checkpoint(Callback):
```
[the remark this answers is not in the document]
Comment (lihuoran) [reply maro/rl/workflows/callback.py:90]: I had a proposal about how to support early-stopping and other similar requirements. Will make a PR (based on this PR) to add it.

File: maro/rl/rollout/env_sampler.py [hunk 4]
```diff
@@ -430,65 +433,71 @@ def sample(
...
+                # Update env and get new states (global & agent)
+                self._step(list(env_action_dict.values()))
+
+                if self._reward_eval_delay is None:
+                    self._calc_reward(cache_element)
+                    self._post_step(cache_element)
+                self._append_cache_element(cache_element)
+                if steps_to_go is not None:
+                    steps_to_go -= 1
+            self._append_cache_element(None)
```
[the remark this answers is not in the document]
Comment (lihuoran) [reply maro/rl/rollout/env_sampler.py:481]: Fixed

File: maro/rl/workflows/callback.py [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+        if ep % self._interval == 0:
+            training_manager.save(os.path.join(self._path, str(ep)))
+            logger.info(f"[Episode {ep}] All trainer states saved under {self._path}")
+
+
+class MetricsRecorder(Callback):
+    def __init__(self, path: str) -> None:
+        super(MetricsRecorder, self).__init__()
+
+        self._full_metrics: Dict[int, dict] = {}
```
[the remark this answers is not in the document]
Comment (lihuoran) [reply maro/rl/workflows/callback.py:113]: Pandas action here (sort, make dataframe, dump) are very light compared to rollout & training. Collecting & dumping metrics from scratch each time will leave flexibility to users (for example, overwrite old information). I personally think it's worth the cost to retain flexibility here.

**Review by Jinyu-W — APPROVED** [review]

**Pull request merged**